src/javascript/expression.rs

```diff
@@ -113,10 +115,66 @@ impl Generator {
     }
 }
 
+fn int(value: &str) -> Document<'_> {
+    value.to_doc()
+}
+
+fn float(value: &str) -> Document<'_> {
+    value.to_doc()
+}
+
 fn string(value: &str) -> Document<'_> {
     value.to_doc().surround("\"", "\"")
 }
 
+fn bin_op<'a>(name: &'a BinOp, left: &'a TypedExpr, right: &'a TypedExpr) -> Output<'a> {
+    // let div_zero = match name {
+    //     BinOp::DivInt | BinOp::ModuloInt => Some("0"),
+    //     BinOp::DivFloat => Some("0.0"),
+    //     _ => None,
+    // };
+    match name {
+        BinOp::And => print_bin_op(left, right, "&&"),
+        BinOp::Or => print_bin_op(left, right, "||"),
+        BinOp::LtInt | BinOp::LtFloat => print_bin_op(left, right, "<"),
+        BinOp::LtEqInt | BinOp::LtEqFloat => print_bin_op(left, right, "<="),
+        // https://dmitripavlutin.com/how-to-compare-objects-in-javascript/
+        // BinOp::Eq => "=:=",
+        // BinOp::NotEq => "/=",
+        BinOp::GtInt | BinOp::GtFloat => print_bin_op(left, right, ">"),
+        BinOp::GtEqInt | BinOp::GtEqFloat => print_bin_op(left, right, ">="),
+        BinOp::AddInt | BinOp::AddFloat => print_bin_op(left, right, "+"),
+        BinOp::SubInt | BinOp::SubFloat => print_bin_op(left, right, "-"),
+        BinOp::MultInt | BinOp::MultFloat => print_bin_op(left, right, "*"),
+        BinOp::DivInt => Ok(Document::String("Math.floor".to_string())
+            .append(print_bin_op(left, right, "/")?.surround("(", ")"))),
+        BinOp::DivFloat => print_bin_op(left, right, "/"),
+        BinOp::ModuloInt => print_bin_op(left, right, "%"),
+        _ => {
+            println!("name: {:?}", name);
+            unimplemented!("binop")
+        }
+    }
+}
+
+fn print_bin_op<'a>(left: &'a TypedExpr, right: &'a TypedExpr, op: &str) -> Output<'a> {
+    let left_expr = match left {
+        TypedExpr::BinOp { .. } => Ok(Generator::compile(left)?.surround("(", ")")),
+        _ => Generator::compile(left),
```

Comment:
@lpil is there some other way to use the `expression` function that you were intending?
I want to re use it here because `left` (and `right`) are terms are `TypedExpr` however using `Generator::compile` now mean all my tests are failing with a return on each line.

Reply:
The same generator is intended to be used for the entire module (as it can carry any state we need such as whether it is in the tail postion). This is failing because the state is being discarded and reset with each new generator so it is back to being in the tail position state.

These functions should be methods on the generator, like in the formatter with the printer.

i.e.

```rust
fn bin_op<'a>(&mut self, left: &'a TypedExpr, right: &'a TypedExpr, op: &str) -> Output<'a> {
    let left_expr = match left {
        TypedExpr::BinOp { .. } => Ok(self.expression(left)?.surround("(", ")")),
// ...
```

Reply:
Ok, I think I follow. Getting a bit OO'y with all this use of `self` not really an approach I think about any more but looks like it makes sense in rust land.

Bit busy this week so might not get round to it till the weekend.

Reply:
Fab, thank you